### DISCUSSION TITLE
谈谈 MVX 中的 Model · /mvx-model

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#258] [Announcements]
https://draveness.me/mvx-model

**Comment by chencanxin** [comment]
get

**Comment by vinnyxiong** [comment]
👍

**Comment by WeMadeCode** [comment]
学习了

**Comment by GTMYang** [comment]
什么是元编程？

**Comment by draveness** [comment]
> @GTMYang
> 什么是元编程？

可以看一下这个 [谈元编程与表达能力](https://draveness.me/metaprogramming)

**Comment by ivaneye** [comment]
博主可能对MVX模式中的Model的理解有些偏差。
MVX中的Model实际上是个泛指，用于给View提供数据的，包含了数据模型和业务逻辑。（可以参考wiki）
博主文中所提到的Model层，Service层等，实际上是分层(Layer)架构中的中的Layer

**Comment by draveness** [comment]
> @ivaneye
> 博主可能对MVX模式中的Model的理解有些偏差。
> MVX中的Model实际上是个泛指，用于给View提供数据的，包含了数据模型和业务逻辑。（可以参考wiki）
> 博主文中所提到的Model层，Service层等，实际上是分层(Layer)架构中的中的Layer

有出处么？想了解一下，我在文章中提到了如下的内容：

>  Rails 中的 Model 层主要承担着以下两大职责：
>
> 使用数据库存储并管理 Web 应用的数据；
> 包含 Web 应用所有的业务逻辑；

和这里说的『数据模型和业务逻辑』有什么偏差么？

**Comment by ivaneye** [comment]
> @draveness
> > @ivaneye
> > 博主可能对MVX模式中的Model的理解有些偏差。
> > MVX中的Model实际上是个泛指，用于给View提供数据的，包含了数据模型和业务逻辑。（可以参考wiki）
> > 博主文中所提到的Model层，Service层等，实际上是分层(Layer)架构中的中的Layer
> 
> 有出处么？想了解一下，我在文章中提到了如下的内容：
> 
> >  Rails 中的 Model 层主要承担着以下两大职责：
> >
> > 使用数据库存储并管理 Web 应用的数据；
> > 包含 Web 应用所有的业务逻辑；
> 
> 和这里说的『数据模型和业务逻辑』有什么偏差么？

博主把MVC架构模式，CS架构风格和LAYER架构风格结合看下就能找到区别了。
rails中的activerecord是个充血模型，包含了数据和业务逻辑，实际上model还有贫血模型，失血模型和胀血模型。
以贫血模型来说，model中只有数据，而没有业务逻辑，这时候就需要另外两层来分别处理业务和持久化，就是service层和DAO层(或者respository层)。
如果是MVC架构的话，model应该是包含了(service layer，model layer和dao layer，后面的持久层也属于model)。
再说controller，博主在另一篇里说到了springmvc的实现，springmvc使用了前端控制器模式，也就是那个dispatcherservlet。对应到MVC，controller指的是从dispatcherservlet到spring的controller层(个人认为叫handler不会引起误解)，dispatcherservlet统一接收请求，分发到各个handler，handler是对各个请求的处理(解构参数，调用后面的MODEL，返回哪个页面)，dispatcherservlet根据每个handler的返回，选择合适的模板进行渲染，返回渲染出来的view。
手机码字太累了～(￣▽￣～)~

**Comment by draveness** [comment]
> @ivaneye
> > @draveness
> > > @ivaneye
> > > 博主可能对MVX模式中的Model的理解有些偏差。
> > > MVX中的Model实际上是个泛指，用于给View提供数据的，包含了数据模型和业务逻辑。（可以参考wiki）
> > > 博主文中所提到的Model层，Service层等，实际上是分层(Layer)架构中的中的Layer
> > 
> > 有出处么？想了解一下，我在文章中提到了如下的内容：
> > 
> > >  Rails 中的 Model 层主要承担着以下两大职责：
> > >
> > > 使用数据库存储并管理 Web 应用的数据；
> > > 包含 Web 应用所有的业务逻辑；
> > 
> > 和这里说的『数据模型和业务逻辑』有什么偏差么？
> 
> 博主把MVC架构模式，CS架构风格和LAYER架构风格结合看下就能找到区别了。
> rails中的activerecord是个充血模型，包含了数据和业务逻辑，实际上model还有贫血模型，失血模型和胀血模型。
> 以贫血模型来说，model中只有数据，而没有业务逻辑，这时候就需要另外两层来分别处理业务和持久化，就是service层和DAO层(或者respository层)。
> 如果是MVC架构的话，model应该是包含了(service layer，model layer和dao layer，后面的持久层也属于model)。
> 再说controller，博主在另一篇里说到了springmvc的实现，springmvc使用了前端控制器模式，也就是那个dispatcherservlet。对应到MVC，controller指的是从dispatcherservlet到spring的controller层(个人认为叫handler不会引起误解)，dispatcherservlet统一接收请求，分发到各个handler，handler是对各个请求的处理(解构参数，调用后面的MODEL，返回哪个页面)，dispatcherservlet根据每个handler的返回，选择合适的模板进行渲染，返回渲染出来的view。
> 手机码字太累了～(￣▽￣～)~

1. iOS/Android 的 MVC 中的 Model 基本都是没有业务逻辑的失血模型、贫血模型；Rails 是包含业务逻辑的充血模型
2. 在另一篇文章介绍的 Controller，同时给出了 AST.NET、Spring MVC 还有 iOS（客户端）对于 MVC 的不同实现

文章中从来没有提到过 X 血模型这一概念，最开始介绍  Model 层最开始介绍的是 iOS 中常见 Model 层是什么样的 --- 只包含数据结构，随后以 Rails 为例介绍了其中包含数据结构和大多数业务逻辑的 Model 层，最后介绍的是在 iOS 中什么样的 Model 层是作者认为比较合理的。

我在之前的编程经历中，只在 Java 的生态环境中体会到了非常多复杂的概念，比如 DAO、Repository、POJO 等等， 但是其他主要用于后端的语言和框架没有这么多复杂的概念，不同的生态由于语言和设计理念的差异对于 MVC 的实现本身上就会有不同，我不认为 MVC 是 100% 确定的设计模式，它只是在某些方面不可变，其他的地方留给工程师一些灵活的余地，让复杂的软件能够更快的开发、交付和维护。

您能提出这样的问题，我还是比较开心的，对于相关问题的讨论和交流，我也非常愿意，不过我还是没理解这里提到的关于 Model 和 Controller 的解释有什么问题，您想表达的是什么，是我这里对于 Model 和 Controller 的解释有问题么，谢谢。

**Comment by ivaneye** [comment]
> @draveness
> > @ivaneye
> > > @draveness
> > > > @ivaneye
> > > > 博主可能对MVX模式中的Model的理解有些偏差。
> > > > MVX中的Model实际上是个泛指，用于给View提供数据的，包含了数据模型和业务逻辑。（可以参考wiki）
> > > > 博主文中所提到的Model层，Service层等，实际上是分层(Layer)架构中的中的Layer
> > > 
> > > 有出处么？想了解一下，我在文章中提到了如下的内容：
> > > 
> > > >  Rails 中的 Model 层主要承担着以下两大职责：
> > > >
> > > > 使用数据库存储并管理 Web 应用的数据；
> > > > 包含 Web 应用所有的业务逻辑；
> > > 
> > > 和这里说的『数据模型和业务逻辑』有什么偏差么？
> > 
> > 博主把MVC架构模式，CS架构风格和LAYER架构风格结合看下就能找到区别了。
> > rails中的activerecord是个充血模型，包含了数据和业务逻辑，实际上model还有贫血模型，失血模型和胀血模型。
> > 以贫血模型来说，model中只有数据，而没有业务逻辑，这时候就需要另外两层来分别处理业务和持久化，就是service层和DAO层(或者respository层)。
> > 如果是MVC架构的话，model应该是包含了(service layer，model layer和dao layer，后面的持久层也属于model)。
> > 再说controller，博主在另一篇里说到了springmvc的实现，springmvc使用了前端控制器模式，也就是那个dispatcherservlet。对应到MVC，controller指的是从dispatcherservlet到spring的controller层(个人认为叫handler不会引起误解)，dispatcherservlet统一接收请求，分发到各个handler，handler是对各个请求的处理(解构参数，调用后面的MODEL，返回哪个页面)，dispatcherservlet根据每个handler的返回，选择合适的模板进行渲染，返回渲染出来的view。
> > 手机码字太累了～(￣▽￣～)~
> 
> 1. iOS/Android 的 MVC 中的 Model 基本都是没有业务逻辑的失血模型、贫血模型；Rails 是包含业务逻辑的充血模型
> 2. 在另一篇文章介绍的 Controller，同时给出了 AST.NET、Spring MVC 还有 iOS（客户端）对于 MVC 的不同实现
> 
> 文章中从来没有提到过 X 血模型这一概念，最开始介绍  Model 层最开始介绍的是 iOS 中常见 Model 层是什么样的 --- 只包含数据结构，随后以 Rails 为例介绍了其中包含数据结构和大多数业务逻辑的 Model 层，最后介绍的是在 iOS 中什么样的 Model 层是作者认为比较合理的。
> 
> 我在之前的编程经历中，只在 Java 的生态环境中体会到了非常多复杂的概念，比如 DAO、Repository、POJO 等等， 但是其他主要用于后端的语言和框架没有这么多复杂的概念，不同的生态由于语言和设计理念的差异对于 MVC 的实现本身上就会有不同，我不认为 MVC 是 100% 确定的设计模式，它只是在某些方面不可变，其他的地方留给工程师一些灵活的余地，让复杂的软件能够更快的开发、交付和维护。
> 
> 您能提出这样的问题，我还是比较开心的，对于相关问题的讨论和交流，我也非常愿意，不过我还是没理解这里提到的关于 Model 和 Controller 的解释有什么问题，您想表达的是什么，是我这里对于 Model 和 Controller 的解释有问题么，谢谢。

我主要做的就是「复杂」的Java的。。。。
在文中，我理解的博主的意思是：文中的「Model层」指的是MVC中的Model！是这个意思吧？
如果不是那就是我没有理解博主的意思！

如果是，那我问一个问题！
博主在 Model 层 += Service 层？  这一节里，说道：在IOS里「Model层」只包含数据模型，「Service层」负责请求！
那请问，这里的「Service层」属于MVC里的哪个？是「Model」吗？
如果是就继续（我理解博主的意思应该是属于「Model」，因为后面「Service层」被整合到了「Model层」里去了）！
如果是这样的话，那「Model」是不是应该等于「Model层」+「Service层」？而不是博主说的「Model层」？

「Model层」指的应该是分层架构里的Layer。MVC将系统分成了Model，Controller和View，在Model内又使用了分层架构，进行了分层，里面可能就只有一个「Model层」，也可能有「Model层」，「Service层」，甚至更多层。。。。。

不知道这样表达，楼主能否理解？我尽力了～～～如果我理解有偏差，欢迎楼主指正～

**Comment by draveness** [comment]
> @ivaneye
> > @draveness
> > > @ivaneye
> > > > @draveness
> > > > > @ivaneye
> > > > > 博主可能对MVX模式中的Model的理解有些偏差。
> > > > > MVX中的Model实际上是个泛指，用于给View提供数据的，包含了数据模型和业务逻辑。（可以参考wiki）
> > > > > 博主文中所提到的Model层，Service层等，实际上是分层(Layer)架构中的中的Layer
> > > > 
> > > > 有出处么？想了解一下，我在文章中提到了如下的内容：
> > > > 
> > > > >  Rails 中的 Model 层主要承担着以下两大职责：
> > > > >
> > > > > 使用数据库存储并管理 Web 应用的数据；
> > > > > 包含 Web 应用所有的业务逻辑；
> > > > 
> > > > 和这里说的『数据模型和业务逻辑』有什么偏差么？
> > > 
> > > 博主把MVC架构模式，CS架构风格和LAYER架构风格结合看下就能找到区别了。
> > > rails中的activerecord是个充血模型，包含了数据和业务逻辑，实际上model还有贫血模型，失血模型和胀血模型。
> > > 以贫血模型来说，model中只有数据，而没有业务逻辑，这时候就需要另外两层来分别处理业务和持久化，就是service层和DAO层(或者respository层)。
> > > 如果是MVC架构的话，model应该是包含了(service layer，model layer和dao layer，后面的持久层也属于model)。
> > > 再说controller，博主在另一篇里说到了springmvc的实现，springmvc使用了前端控制器模式，也就是那个dispatcherservlet。对应到MVC，controller指的是从dispatcherservlet到spring的controller层(个人认为叫handler不会引起误解)，dispatcherservlet统一接收请求，分发到各个handler，handler是对各个请求的处理(解构参数，调用后面的MODEL，返回哪个页面)，dispatcherservlet根据每个handler的返回，选择合适的模板进行渲染，返回渲染出来的view。
> > > 手机码字太累了～(￣▽￣～)~
> > 
> > 1. iOS/Android 的 MVC 中的 Model 基本都是没有业务逻辑的失血模型、贫血模型；Rails 是包含业务逻辑的充血模型
> > 2. 在另一篇文章介绍的 Controller，同时给出了 AST.NET、Spring MVC 还有 iOS（客户端）对于 MVC 的不同实现
> > 
> > 文章中从来没有提到过 X 血模型这一概念，最开始介绍  Model 层最开始介绍的是 iOS 中常见 Model 层是什么样的 --- 只包含数据结构，随后以 Rails 为例介绍了其中包含数据结构和大多数业务逻辑的 Model 层，最后介绍的是在 iOS 中什么样的 Model 层是作者认为比较合理的。
> > 
> > 我在之前的编程经历中，只在 Java 的生态环境中体会到了非常多复杂的概念，比如 DAO、Repository、POJO 等等， 但是其他主要用于后端的语言和框架没有这么多复杂的概念，不同的生态由于语言和设计理念的差异对于 MVC 的实现本身上就会有不同，我不认为 MVC 是 100% 确定的设计模式，它只是在某些方面不可变，其他的地方留给工程师一些灵活的余地，让复杂的软件能够更快的开发、交付和维护。
> > 
> > 您能提出这样的问题，我还是比较开心的，对于相关问题的讨论和交流，我也非常愿意，不过我还是没理解这里提到的关于 Model 和 Controller 的解释有什么问题，您想表达的是什么，是我这里对于 Model 和 Controller 的解释有问题么，谢谢。
> 
> 我主要做的就是「复杂」的Java的。。。。
> 在文中，我理解的博主的意思是：文中的「Model层」指的是MVC中的Model！是这个意思吧？
> 如果不是那就是我没有理解博主的意思！
> 
> 如果是，那我问一个问题！
> 博主在 Model 层 += Service 层？  这一节里，说道：在IOS里「Model层」只包含数据模型，「Service层」负责请求！
> 那请问，这里的「Service层」属于MVC里的哪个？是「Model」吗？
> 如果是就继续（我理解博主的意思应该是属于「Model」，因为后面「Service层」被整合到了「Model层」里去了）！
> 如果是这样的话，那「Model」是不是应该等于「Model层」+「Service层」？而不是博主说的「Model层」？
> 
> 「Model层」指的应该是分层架构里的Layer。MVC将系统分成了Model，Controller和View，在Model内又使用了分层架构，进行了分层，里面可能就只有一个「Model层」，也可能有「Model层」，「Service层」，甚至更多层。。。。。
> 
> 不知道这样表达，楼主能否理解？我尽力了～～～如果我理解有偏差，欢迎楼主指正～

不是特别理解，iOS 里的 Service 其实就是获取数据的一种方式，也可以理解成业务逻辑，和服务端不一样的是，Service 几乎是客户端获取 Model 源数据的唯一方式，后端的 Model 后面一般接的是 DB，客户端的 Model 后面接的就是服务端的 API，这里说的 Service 其实和服务端说的 Service 不大一样。

我不知道问题是不是由于客户端和服务端对于 Service 的定义不同，造成这里的沟通和理解问题？

**Comment by ivaneye** [comment]
> @draveness
> > @ivaneye
> > > @draveness
> > > > @ivaneye
> > > > > @draveness
> > > > > > @ivaneye
> > > > > > 博主可能对MVX模式中的Model的理解有些偏差。
> > > > > > MVX中的Model实际上是个泛指，用于给View提供数据的，包含了数据模型和业务逻辑。（可以参考wiki）
> > > > > > 博主文中所提到的Model层，Service层等，实际上是分层(Layer)架构中的中的Layer
> > > > > 
> > > > > 有出处么？想了解一下，我在文章中提到了如下的内容：
> > > > > 
> > > > > >  Rails 中的 Model 层主要承担着以下两大职责：
> > > > > >
> > > > > > 使用数据库存储并管理 Web 应用的数据；
> > > > > > 包含 Web 应用所有的业务逻辑；
> > > > > 
> > > > > 和这里说的『数据模型和业务逻辑』有什么偏差么？
> > > > 
> > > > 博主把MVC架构模式，CS架构风格和LAYER架构风格结合看下就能找到区别了。
> > > > rails中的activerecord是个充血模型，包含了数据和业务逻辑，实际上model还有贫血模型，失血模型和胀血模型。
> > > > 以贫血模型来说，model中只有数据，而没有业务逻辑，这时候就需要另外两层来分别处理业务和持久化，就是service层和DAO层(或者respository层)。
> > > > 如果是MVC架构的话，model应该是包含了(service layer，model layer和dao layer，后面的持久层也属于model)。
> > > > 再说controller，博主在另一篇里说到了springmvc的实现，springmvc使用了前端控制器模式，也就是那个dispatcherservlet。对应到MVC，controller指的是从dispatcherservlet到spring的controller层(个人认为叫handler不会引起误解)，dispatcherservlet统一接收请求，分发到各个handler，handler是对各个请求的处理(解构参数，调用后面的MODEL，返回哪个页面)，dispatcherservlet根据每个handler的返回，选择合适的模板进行渲染，返回渲染出来的view。
> > > > 手机码字太累了～(￣▽￣～)~
> > > 
> > > 1. iOS/Android 的 MVC 中的 Model 基本都是没有业务逻辑的失血模型、贫血模型；Rails 是包含业务逻辑的充血模型
> > > 2. 在另一篇文章介绍的 Controller，同时给出了 AST.NET、Spring MVC 还有 iOS（客户端）对于 MVC 的不同实现
> > > 
> > > 文章中从来没有提到过 X 血模型这一概念，最开始介绍  Model 层最开始介绍的是 iOS 中常见 Model 层是什么样的 --- 只包含数据结构，随后以 Rails 为例介绍了其中包含数据结构和大多数业务逻辑的 Model 层，最后介绍的是在 iOS 中什么样的 Model 层是作者认为比较合理的。
> > > 
> > > 我在之前的编程经历中，只在 Java 的生态环境中体会到了非常多复杂的概念，比如 DAO、Repository、POJO 等等， 但是其他主要用于后端的语言和框架没有这么多复杂的概念，不同的生态由于语言和设计理念的差异对于 MVC 的实现本身上就会有不同，我不认为 MVC 是 100% 确定的设计模式，它只是在某些方面不可变，其他的地方留给工程师一些灵活的余地，让复杂的软件能够更快的开发、交付和维护。
> > > 
> > > 您能提出这样的问题，我还是比较开心的，对于相关问题的讨论和交流，我也非常愿意，不过我还是没理解这里提到的关于 Model 和 Controller 的解释有什么问题，您想表达的是什么，是我这里对于 Model 和 Controller 的解释有问题么，谢谢。
> > 
> > 我主要做的就是「复杂」的Java的。。。。
> > 在文中，我理解的博主的意思是：文中的「Model层」指的是MVC中的Model！是这个意思吧？
> > 如果不是那就是我没有理解博主的意思！
> > 
> > 如果是，那我问一个问题！
> > 博主在 Model 层 += Service 层？  这一节里，说道：在IOS里「Model层」只包含数据模型，「Service层」负责请求！
> > 那请问，这里的「Service层」属于MVC里的哪个？是「Model」吗？
> > 如果是就继续（我理解博主的意思应该是属于「Model」，因为后面「Service层」被整合到了「Model层」里去了）！
> > 如果是这样的话，那「Model」是不是应该等于「Model层」+「Service层」？而不是博主说的「Model层」？
> > 
> > 「Model层」指的应该是分层架构里的Layer。MVC将系统分成了Model，Controller和View，在Model内又使用了分层架构，进行了分层，里面可能就只有一个「Model层」，也可能有「Model层」，「Service层」，甚至更多层。。。。。
> > 
> > 不知道这样表达，楼主能否理解？我尽力了～～～如果我理解有偏差，欢迎楼主指正～
> 
> 不是特别理解，iOS 里的 Service 其实就是获取数据的一种方式，也可以理解成业务逻辑，和服务端不一样的是，Service 几乎是客户端获取 Model 源数据的唯一方式，后端的 Model 后面一般接的是 DB，客户端的 Model 后面接的就是服务端的 API，这里说的 Service 其实和服务端说的 Service 不大一样。
> 
> 我不知道问题是不是由于客户端和服务端对于 Service 的定义不同，造成这里的沟通和理解问题？

和定义没有关系！我是说你所说的「Model层」应该是MVC中的「Model」的一部分，而不是「Model」本身！
MVC中的「Model」可能还包含「Service层」等其它层！

「Model是包含数据和业务逻辑的」，这句你认同吧？

那前端MVC的Model应该 = 「Model层」（这是数据） + 「Service层」 + 「后端服务」（这是业务）
后端MVC中的Model应该 = 「Model层」（这是数据） + 「Service层」+  「一堆乱七八糟的层」（这是业务）

「Model层」是「Model」的子集！

**Comment by HuangYanQi** [comment]
对于OC相比Swift提高编程效率，不敢苟同。

**Comment by doubleLLL3** [comment]
看完MVC感觉悟了，看到MVP、MVVM，感觉不是很清楚它们的区别了，后面实践的时候再体会吧。:smile:
